### PR TITLE
 pages: add sections about local support for bindings in wrangler.toml

### PR DESCRIPTION
**Pull request by admah**
Updated local bindings sections for: kv, r2, d1, and service bindings since those are currently supported.

### CODE REVIEW

**Comment by Cherry** [comment]
Pages using `wrangler.toml` locally but not when you publish _really_ confuses people. We see it every single week in the Discord.

https://github.com/cloudflare/workers-sdk/issues/3757

It seems a little strange to me to encourage this behaviour until Pages actually has full support for configuration in some capacity.

**Comment by admah** [comment]
@Cherry the goal of adding this would be to clear up that confusion for users by stating explicitly what is supported now. We are working towards support for Pages file-based configuration so we're not going to remove what is currently exposed. As we work towards that, we are also using this functionality to support better framework development like in [`next-on-pages`](https://github.com/cloudflare/next-on-pages/pull/648) and via the [getBindingsProxy API](https://developers.cloudflare.com/workers/wrangler/api/#getbindingsproxy).

**Comment by Cherry** [comment]
I left [a comment](https://github.com/cloudflare/workers-sdk/issues/3757#issuecomment-1917332954) on the `workers-sdk` PR, but I'm personally not sure this reduces confusion. 🤷‍♀️ Leaning into what has been considered a bug for years, with everyone being told that Pages does not support `wrangler.toml`, seems odd to me.

Especially if folks are going to have to update their configs again when Pages does support configuration.

File: content/pages/functions/bindings.md [hunk 1]
```diff
@@ -306,6 +336,14 @@ export const onRequest: PagesFunction<Env> = async (context) => {
 
 To interact with a [service binding](/workers/configuration/bindings/about-service-bindings/) while developing locally, run the Worker you want to bind to via `wrangler dev` and in parallel, run `wrangler pages dev` with `--service <BINDING_NAME>=<SCRIPT_NAME>` where `SCRIPT_NAME` indicates the name of the Worker. For example, if your Worker is called `my-worker`, connect with this Worker by running it via `npx wrangler dev` (in the Worker's directory) alongside `npx wrangler pages dev <OUTPUT_DIR> --service MY_SERVICE=my-worker` (in the Pages' directory). Interact with this binding by using `context.env` (for example, `context.env.MY_SERVICE`).
 
+Alternatively, you can interact with a service binding locally via a `wrangler.toml` file. See [Wrangler configuration](/workers/wrangler/configuration/#service-bindings) for more information.
+
+{{<Aside type="note">}}
+
+`wrangler.toml` is currently **only** used for local development. Bindings specified in it are not available remotely.
```
Comment (jkonowitch) [inline content/pages/functions/bindings.md]: These callouts are very helpful

**Comment by jkonowitch** [comment]
> We are working towards support for Pages file-based configuration

Just to make sure I understand, is the plan to support `wrangler.toml` configuration for deploys on-par with how Workers work?

File: content/pages/functions/bindings.md [hunk 2]
```diff
@@ -55,6 +55,14 @@ export const onRequest: PagesFunction<Env> = async (context) => {
 
 While developing locally, interact with your KV namespace by adding `-k <BINDING_NAME>` or `--kv=<BINDING_NAME>` to your run command. For example, if your namespace is bound to `TODO_LIST`, access the KV namespace in your local dev by running `npx wrangler pages dev <OUTPUT_DIR> --kv=TODO_LIST`. The data from this namespace can be accessed using `context.env.TODO_LIST`.
 
+Alternatively, you can interact with a KV namespace locally via a `wrangler.toml` file. See [Wrangler configuration](/workers/wrangler/configuration/#kv-namespaces) for more information.
```
Comment (deadlypants1973) [inline content/pages/functions/bindings.md]:
```suggestion
Alternatively, you can interact with a KV namespace locally via a `wrangler.toml` file. Refer to [Wrangler configuration](/workers/wrangler/configuration/#kv-namespaces) for more information.
```

File: content/pages/functions/bindings.md [hunk 3]
```diff
@@ -147,7 +155,21 @@ export const onRequest: PagesFunction<Env> = async (context) => {
...
-While developing locally, interact with an R2 bucket by adding `--r2=<BINDING_NAME>` to your run command. For example, if your bucket is bound to `BUCKET`, access this bucket in local dev by running `npx wrangler pages dev <OUTPUT_DIR> --r2=BUCKET`. Interact with this binding by using `context.env` (for example, `context.env.BUCKET`).
+While developing locally, interact with an R2 bucket by adding `--r2=<BINDING_NAME>` to your run command.
+
+{{<Aside type="note">}}
+By default, `wrangler dev` automatically persists data.
+{{</Aside>}}
+
+If your bucket is bound to `BUCKET`, access this bucket in local dev by running `npx wrangler pages dev <OUTPUT_DIR> --r2=BUCKET`. Interact with this binding by using `context.env` (for example, `context.env.BUCKET`).
+
+Alternatively, you can interact with an R2 bucket locally via a `wrangler.toml` file. See [Wrangler configuration](/workers/wrangler/configuration/#r2-buckets) for more information.
```
Comment (deadlypants1973) [inline content/pages/functions/bindings.md]:
```suggestion
Alternatively, you can interact with an R2 bucket locally via a `wrangler.toml` file. Refer to [Wrangler configuration](/workers/wrangler/configuration/#r2-buckets) for more information.
```

File: content/pages/functions/bindings.md [hunk 4]
```diff
@@ -207,6 +229,14 @@ Specifically:
 
 Refer to the [D1 client API documentation](/d1/how-to/query-databases/) for the API methods available on your D1 binding.
 
+Alternatively, you can interact with a D1 database locally via a `wrangler.toml` file. See [Wrangler configuration](/workers/wrangler/configuration/#d1-databases) for more information.
```
Comment (deadlypants1973) [inline content/pages/functions/bindings.md]:
```suggestion
Alternatively, you can interact with a D1 database locally via a `wrangler.toml` file. Refer to [Wrangler configuration](/workers/wrangler/configuration/#d1-databases) for more information.
```

File: content/pages/functions/bindings.md [hunk 5]
```diff
@@ -306,6 +336,14 @@ export const onRequest: PagesFunction<Env> = async (context) => {
 
 To interact with a [service binding](/workers/configuration/bindings/about-service-bindings/) while developing locally, run the Worker you want to bind to via `wrangler dev` and in parallel, run `wrangler pages dev` with `--service <BINDING_NAME>=<SCRIPT_NAME>` where `SCRIPT_NAME` indicates the name of the Worker. For example, if your Worker is called `my-worker`, connect with this Worker by running it via `npx wrangler dev` (in the Worker's directory) alongside `npx wrangler pages dev <OUTPUT_DIR> --service MY_SERVICE=my-worker` (in the Pages' directory). Interact with this binding by using `context.env` (for example, `context.env.MY_SERVICE`).
 
+Alternatively, you can interact with a service binding locally via a `wrangler.toml` file. See [Wrangler configuration](/workers/wrangler/configuration/#service-bindings) for more information.
```
Comment (deadlypants1973) [inline content/pages/functions/bindings.md]:
```suggestion
Alternatively, you can interact with a service binding locally via a `wrangler.toml` file. Refer to [Wrangler configuration](/workers/wrangler/configuration/#service-bindings) for more information.
```

**Review by deadlypants1973 — APPROVED** [review]

**Comment by Cherry** [comment]
I think it would be extremely beneficial to hold this PR until it's made very clear in `wrangler pages deploy` commands and Pages CI that publishing a project with bindings in your `wrangler.toml` won't actually make them available in production. This is a serious confusion and pain point people hit constantly with Pages commands, with the `wrangler.toml` local functionality being described as a bug for years.

This is made especially and even more confusing when these new docs talk about Pages, link to wrangler configuration docs that detail how this _does_ work when deploying (the linked docs pages are tailored for Workers), but then doesn't and no warnings are presented to the user. Most users struggle to understand the nuances between Workers and Pages right now, and I personally think this only adds to the confusion.

**Pull request merged**